src/lib.rs

```diff
@@ -36,10 +36,36 @@ static ALLOCATOR: LibcAlloc = LibcAlloc;
 unsafe impl GlobalAlloc for LibcAlloc {
     #[inline]
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-        libc::memalign(
-            layout.align().max(core::mem::size_of::<usize>()),
-            layout.size(),
-        ) as *mut u8
+        let align = layout.align().max(core::mem::size_of::<usize>());
+        let size = layout.size();
+
+        #[cfg(not(target_os = "macos"))]
+        {
+            libc::memalign(align, size) as *mut u8
+        }
+
+        #[cfg(target_os = "macos")]
+        {
+            // Initialize ptr as a null pointer of the return type, *mut u8.
+            let mut ptr: *mut u8 = core::ptr::null_mut();
+
+            let result = libc::posix_memalign(
+                // Convert the reference to ptr (&mut ptr) into a mutable raw pointer (*mut _).
+                // The underscore _ allows Rust to infer the specific type, which in this case
+                // becomes a raw pointer to a mutable pointer (*mut *mut u8).
+                &mut ptr as *mut _
+                // Further cast the raw pointer to *mut *mut c_void, the type required by
+                // posix_memalign. This casting is necessary for FFI compatibility, as C functions
+                // like posix_memalign do not recognize Rust-specific pointer types.
+                as *mut *mut c_void,
+                align,
+                size,
+            );
+            match result {
+                0 => ptr,                   // Allocation successful
+                _ => core::ptr::null_mut(), // Allocation failed
+            }
```

Comment:
While I appreciate the attention to detail in your comments, and your focus on making sure this code is indeed correct, I think this more succinct formulation should suffice

```suggestion
            let mut ptr: *mut u8 = core::ptr::null_mut();
            let res = libc::posix_memalign(
                (&mut ptr as *mut *mut u8).cast::<*mut c_void>(),
                align,
                size,
            );
            if res != 0 {
                return core::ptr::null_mut();
            }
            ptr
```

Please double check that this code actually works correctly though - I don't have a macOS system to test with myself :)